### PR TITLE
KFSPTS-16511: New integration to create csv file of closed accounts for WorkDay. Code has ability to create initial seed file as well as incremental file based on KFS system parameter.

### PR DESCRIPTION
**Pull request by nkimble**

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/coa/batch/CreateClosedLaborAccountsCsvByPreviousDaysStep.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.util.Date;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.kuali.kfs.sys.batch.AbstractStep;
+
+import edu.cornell.kfs.coa.batch.service.CreateClosedLaborAccountsCsvByPreviousDaysService;
+
+public class CreateClosedLaborAccountsCsvByPreviousDaysStep extends AbstractStep {
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/CreateClosedLaborAccountsCsvByPreviousDaysStep.java]: I think this step class name is too wordy, How about just naming the class "CreateClosedLaborAccountsCsvStep"

File: src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchConstants.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package edu.cornell.kfs.coa.batch;
+
+import java.sql.Date;
+
+public class CuCoaBatchConstants {
+    
+    public static final String CSV_FILE_EXTENSION = ".csv";
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchConstants.java]: Can we use the constant for CSV file extension in KFSConstants?

File: src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchConstants.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    
+    public static class ClosedLaborAccountsFileCreationConstants {
+    
+        public static final int PARAMETER_LABOR_ACCOUNTS_CLOSED_OVER_PAST_DAYS_NOT_SET = -1;
+        public static final int PARAMETER_SET_TO_CREATE_FULL_SEED_FILE = 0;
+        
+        public static final Date SEED_FILE_DEFAULT_FROM_DATE = Date.valueOf("2013-03-01");
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchConstants.java]: I am wondering if we should move this value to a parameter as well.  The use case I am thinking is workday wants a full extract again, but only back to a certain date.

File: src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchParameterConstants.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package edu.cornell.kfs.coa.batch;
+
+public class CuCoaBatchParameterConstants {
+    
+    public static final String CREATE_CLOSED_LABOR_ACCOUNTS_CSV_BY_PREVIOUS_DAYS_STEP = "createClosedLaborAccountsCsvByPreviousDaysStep";
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchParameterConstants.java]: assuming the step name gets shortened, this will need updated as well

File: src/main/java/edu/cornell/kfs/coa/batch/businessobject/LaborClosedAccount.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    public Date getAccountClosedDate() {
+        return accountClosedDate;
+    }
+
+    public void setAccountClosedDate(Date accountClosedDate) {
+        this.accountClosedDate = accountClosedDate;
+    }
+    
+    public String toCsvString() {
+        SimpleDateFormat sdf = new SimpleDateFormat("MM/dd/yyyy", Locale.US);
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/businessobject/LaborClosedAccount.java]: "MM/DD/yyyy" is defined as a constant in KFSConstants

File: src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    
+    private static final String ACCT_FIN_COA_CD = "ACCT_FIN_COA_CD";
+    private static final String ACCT_ACCOUNT_NBR = "ACCT_ACCOUNT_NBR";
+    private static final String SUB_SUB_ACCT_NBR = "SUB_SUB_ACCT_NBR";
+    private static final String ACCT_ACCT_CLOSED_IND = "ACCT_ACCT_CLOSED_IND";
+    private static final String ACCTX_ACCT_CLOSED_DT = "ACCTX_ACCT_CLOSED_DT";
+    
+    @Override
+    public List<LaborClosedAccount> obtainLaborClosedAccountsDataFor(Map<String, Date> dateRange) {
+        List<LaborClosedAccount> closedLaborAccounts = findAllClosedLaborAccountsFor(dateRange);
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java]: Is there a reason to have this function be a one liner, calling the find function?  Why not just have ClosedLaborAccountsByDateRangeDao define findAllClosedLaborAccountsFor and get rid of obtainLaborClosedAccountsDataFor?

File: src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            };
+            return this.getJdbcTemplate().query(buildClosedLaborAccountsForDateRangeSQL(dateRange), mapRow);
+        } catch (Exception e) {
+            LOG.info("findAllClosedLaborAccountsFor Exception: " + e.getMessage());
+            return null;
+        }
+    }
+
+    private String buildClosedLaborAccountsForDateRangeSQL(Map<String, Date> dateRange) {
+        SimpleDateFormat formatter = new SimpleDateFormat("dd-MMM-yy", Locale.US);
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java]: should we add "dd-MMM-yy" as a constant in CUKFSConstants, there are a number of date formats defined there.

File: src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        StringBuilder sqlBuilder = new StringBuilder("SELECT A.FIN_COA_CD AS ACCT_FIN_COA_CD, A.ACCOUNT_NBR AS ACCT_ACCOUNT_NBR, ");
+        sqlBuilder.append("S.SUB_ACCT_NBR AS SUB_SUB_ACCT_NBR, A.ACCT_CLOSED_IND AS ACCT_ACCT_CLOSED_IND, X.ACCT_CLOSED_DT AS ACCTX_ACCT_CLOSED_DT ");
+        sqlBuilder.append("FROM KFS.CA_ACCOUNT_T A ");
+        sqlBuilder.append("LEFT JOIN KFS.CA_ACCOUNT_TX X ON A.FIN_COA_CD = X.FIN_COA_CD AND A.ACCOUNT_NBR = X.ACCOUNT_NBR ");
+        sqlBuilder.append("LEFT JOIN KFS.CA_SUB_ACCT_T S ON A.FIN_COA_CD = S.FIN_COA_CD AND A.ACCOUNT_NBR = S.ACCOUNT_NBR ");
+        sqlBuilder.append("WHERE A.FIN_COA_CD = X.FIN_COA_CD AND A.ACCOUNT_NBR = X.ACCOUNT_NBR AND A.FIN_COA_CD = 'IT' AND A.ACCT_CLOSED_IND = 'Y' ");
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java]: Should we move the chart to a parameter?  Would we ever need to submit accounts from the other charts?

File: src/main/java/edu/cornell/kfs/coa/batch/service/CreateClosedLaborAccountsCsvByPreviousDaysService.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package edu.cornell.kfs.coa.batch.service;
+
+public interface CreateClosedLaborAccountsCsvByPreviousDaysService {
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/service/CreateClosedLaborAccountsCsvByPreviousDaysService.java]: should this be renamed to CreateClosedLaborAccountsCsvService

File: src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvByPreviousDaysServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+import edu.cornell.kfs.coa.batch.CuCoaBatchConstants;
+import edu.cornell.kfs.coa.batch.CuCoaBatchParameterConstants;
+import edu.cornell.kfs.coa.batch.businessobject.LaborClosedAccount;
+import edu.cornell.kfs.coa.batch.dataaccess.ClosedLaborAccountsByDateRangeDao;
+import edu.cornell.kfs.coa.batch.service.CreateClosedLaborAccountsCsvByPreviousDaysService;
+import edu.cornell.kfs.sys.CUKFSConstants;
+import edu.cornell.kfs.sys.CUKFSParameterKeyConstants;
+
+public class CreateClosedLaborAccountsCsvByPreviousDaysServiceImpl implements CreateClosedLaborAccountsCsvByPreviousDaysService {
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvByPreviousDaysServiceImpl.java]: should the be renamed to CreateClosedLaborAccountsCsvServiceImpl?

File: src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvByPreviousDaysServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+    
+    private Integer getLaborAccoutsClosedPastDaysParameter() {
+        String daysPreviousParameterValue = getParameterService().getParameterValueAsString(KFSConstants.CoreModuleNamespaces.CHART, 
+                CUKFSParameterKeyConstants.ALL_COMPONENTS, CuCoaBatchParameterConstants.LABOR_ACCOUNTS_CLOSED_OVER_PAST_DAYS);
+        
+        int daysPreviousInt;
+        try {
+            daysPreviousInt = Integer.parseInt(daysPreviousParameterValue);
+        } catch (NumberFormatException ne) {
+            LOG.error("getLaborAccoutsClosedPastDaysParameter: Namespace KFS-COA system parameter LABOR_ACCOUNTS_CLOSED_OVER_PAST_DAYS is either not set or could not be converted to a number.");
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvByPreviousDaysServiceImpl.java]: I would put the value of daysPreviousParameterValue in the log message

File: src/main/resources/edu/cornell/kfs/coa/cu-spring-coa.xml [hunk 12]
```diff
@@ -204,5 +205,30 @@
             </set>
         </property>
     </bean>
+    
+    <bean id="createClosedLaborAccountsCsvByPreviousDaysJob" parent="unscheduledJobDescriptor">
```
Comment (jhulslander) [inline src/main/resources/edu/cornell/kfs/coa/cu-spring-coa.xml]: How about the job renamed to createClosedLaborAccountsCsvJob

File: src/main/java/edu/cornell/kfs/coa/batch/businessobject/LaborClosedAccount.java (hunk 5 above)
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/coa/batch/businessobject/LaborClosedAccount.java]: Changed.

File: src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java (hunk 6 above)
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java]: This class was modeled after another class that needed to perform processing on the data after the "find" before it was returned to the caller.  A reason to leave it as a current one liner is that it would simplify future maintenance since the "find" would already be fully encapsulated.

File: src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java (hunk 7 above)
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java]: Added new constant.

File: src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvByPreviousDaysServiceImpl.java (hunk 11 above)
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvByPreviousDaysServiceImpl.java]: Done.

File: src/main/resources/edu/cornell/kfs/coa/cu-spring-coa.xml (hunk 12 above)
Comment (nkimble) [reply src/main/resources/edu/cornell/kfs/coa/cu-spring-coa.xml]: Renamed.

File: src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchConstants.java (hunk 3 above)
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchConstants.java]: Moving to a parameter.

File: src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchParameterConstants.java (hunk 4 above)
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchParameterConstants.java]: Done.

**Comment by nkimble** [comment]
I am still validating these code review changes locally but wanted your feedback on a couple of items.

File: src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java (hunk 8 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java]: lets leave as is then

File: src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchConstants.java (hunk 2 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/coa/batch/CuCoaBatchConstants.java]: I spoke with Nancy and we are going to create a CUKFSConstants.FileExtension class add it there

File: src/main/java/edu/cornell/kfs/coa/batch/businessobject/LaborClosedAccount.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    
+    public LaborClosedAccount() {
+        this.chart = null;
+        this.accountNumber = null;
+        this.subAccountNumber = null;
+        this.accountClosedIndicator = null;
+        this.accountClosedDate = null;
+    }
+    
+    public LaborClosedAccount(String chart, String  accountNumber, String subAccountNumber, String accountClosedIndicator, Date accountClosedDate) {
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/businessobject/LaborClosedAccount.java]: I don't think this constructor is used, is it needed?

File: src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+        ensureOutbundDirectoryExists();
+        
+        LOG.info("writeClosedLaborAccountsToCsvFormattedFile: Prior to obtaining data with SQL query for date range. This may take a while.");
+        List<LaborClosedAccount> laborClosedAccountsData = getClosedLaborAccountsByDateRangeDao().obtainLaborClosedAccountsDataFor(fileDataRangeSpecified);
+        
+        LOG.info("writeClosedLaborAccountsToCsvFormattedFile: Have the data. Will now write it out.");
+        writeClosedAccountsToCsvFile(laborClosedAccountsData);
+    }
+    
+    private void ensureOutbundDirectoryExists() {
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java]: I think if you configure the output directory in Spring, it will create the directory for us.  If so, this function may not be needed

File: src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+        FileWriter outputFileWriter;
+        try {
+            outputFileWriter = new FileWriter(outputFile);
+            for (LaborClosedAccount closedAccount : laborClosedAccountsDataList) {
+                outputFileWriter.write(closedAccount.toCsvString());
+                outputFileWriter.write(KFSConstants.NEWLINE);
+            }
+            outputFileWriter.close();
+        } catch (IOException io) {
+            LOG.error("writeClosedAccountsToCsvFile: Caught IOException = " + io.getMessage());
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java]: instead of logging just the message, I would suggest logging the full exception object.  

Also, if the file can't be written, should the job error?  As written, I think the exception is just logged, but the problem isn't percolated back up.

File: src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java (hunk 14 above)
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java]: Confirmed directory is created when originally not present.  Removing method.

File: src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java (hunk 15 above)
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java]: Changing to log io object and bubble up IOException to fail batch job when exception occurs.

File: src/main/java/edu/cornell/kfs/coa/batch/businessobject/LaborClosedAccount.java (hunk 13 above)
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/coa/batch/businessobject/LaborClosedAccount.java]: Removed constructor.

File: src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+        FileWriter outputFileWriter;
+        try {
+            outputFileWriter = new FileWriter(outputFile);
+            for (LaborClosedAccount closedAccount : laborClosedAccountsDataList) {
+                outputFileWriter.write(closedAccount.toCsvString());
+                outputFileWriter.write(KFSConstants.NEWLINE);
+            }
+            outputFileWriter.close();
+        } catch (IOException io) {
+            LOG.error("writeClosedAccountsToCsvFile: Caught IOException = " + io.getStackTrace().toString());
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java]: I think you want it like this:
LOG.error("writeClosedAccountsToCsvFile: Caught IOException", io);

The logger will know it is logging an exception

File: src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java (hunk 6 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/coa/batch/dataaccess/impl/ClosedLaborAccountsByDateRangeDaoJdbc.java]: ok

File: src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java [hunk 17]
```diff
@@ -168,8 +168,13 @@ private void writeClosedLaborAccountsToCsvFormattedFile(Map<String, Date> fileDa
         LOG.info("writeClosedLaborAccountsToCsvFormattedFile: Prior to obtaining data with SQL query for date range. This may take a while.");
         List<LaborClosedAccount> laborClosedAccountsData = getClosedLaborAccountsByDateRangeDao().obtainLaborClosedAccountsDataFor(fileDataRangeSpecified);
         
-        LOG.info("writeClosedLaborAccountsToCsvFormattedFile: Have the data. Will now write it out.");
-        writeClosedAccountsToCsvFile(laborClosedAccountsData);
+        if (ObjectUtils.isNotNull(laborClosedAccountsData) && laborClosedAccountsData.size() > 0) {
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/batch/service/impl/CreateClosedLaborAccountsCsvServiceImpl.java]: This could be simplified to use CollectionUtils.isEmpty
